sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/AsyncCacheNonBlocking.java

```diff
@@ -0,0 +1,206 @@
+     *     Once the refresh task is complete it will be returned to caller.
+     *     If it is a success the value in the cache will be updated. If the refresh task throws an exception the key will be removed from the cache.
+     * </p>
+     *
+     * <p>
+     * If previous initialization function failed - new one will be launched.
+     * </p>
+     *
+     * @param key Key for which to get a value.
+     * @param singleValueInitFunc Initialization function.
+     * @param forceRefresh Force refresh for refreshing the cache
+     * @return Cached value or value returned by initialization function.
+     */
+    public Mono<TValue> getAsync(
+        TKey key,
+        Function<TValue, Mono<TValue>> singleValueInitFunc,
+        Function<TValue, Boolean> forceRefresh) {
+
+        AsyncLazyWithRefresh<TValue> initialLazyValue = values.get(key);
+        if (initialLazyValue != null) {
+            if (logger.isDebugEnabled()) {
+                logger.debug("cache[{}] exists", key);
+            }
+
+            return initialLazyValue.getValueAsync().flatMap(value -> {
+                if(!forceRefresh.apply(value)) {
+                    return Mono.just(value);
+                }
+
+                Mono<TValue> refreshMono = initialLazyValue.createAndWaitForBackgroundRefreshTaskAsync(key, singleValueInitFunc);
+
+                return refreshMono.onErrorResume(
+                    (exception) -> {
+                        if (logger.isDebugEnabled()) {
+                            logger.debug("refresh cache [{}] resulted in error", key, exception);
+                        }
+
+                        if (initialLazyValue.shouldRemoveFromCache()) {
+                            // In some scenarios when a background failure occurs like a 404 the initial cache value should be removed.
+                            if (removeNotFoundFromCacheException((CosmosException)exception)) {
+                                this.remove(key);
+                            }
+                        }
+                        return Mono.error(exception);
+                    }
+                );
+            }).onErrorResume((exception) -> {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("cache[{}] resulted in error", key, exception);
+                }
+                if (initialLazyValue.shouldRemoveFromCache()) {
+                    this.remove(key);
+                }
+                return Mono.error(exception);
+            });
+        }
+
+        if (logger.isDebugEnabled()) {
+            logger.debug("cache[{}] doesn't exist, computing new value", key);
+        }
+        AsyncLazyWithRefresh<TValue> asyncLazyWithRefresh = new AsyncLazyWithRefresh<TValue>(singleValueInitFunc);
+        this.values.putIfAbsent(key, asyncLazyWithRefresh);
+        AsyncLazyWithRefresh<TValue> result = this.values.get(key);
```

Comment:
I think this code path can be simplified to avoid extra get call like this -> 

```
AsyncLazyWithRefresh<TValue> tempResult = this.values.putIfAbsent(key, asyncLazyWithRefresh);
        if (tempResult == null) {
            tempResult = asyncLazyWithRefresh;
        }

        AsyncLazyWithRefresh<TValue> result = tempResult;
```

Reply:
Updated the code.